src/legacy/core_plugins/embeddable_api/public/containers/container.ts

```diff
@@ -0,0 +1,381 @@
+  // The type of embeddable in this panel. Will be used to find the factory in which to
+  // load the embeddable.
+  type: string;
+
+  // Stores input for this embeddable that is specific to this embeddable. Other parts of embeddable input
+  // will be derived from the container's input. Any state in here will override any state derived from
+  // the container.
+  explicitInput: Partial<TEmbeddableInput>;
+}
+export abstract class Container<
+  TChildInput extends Partial<EmbeddableInput> = {},
+  TContainerInput extends ContainerInput = ContainerInput,
+  TContainerOutput extends ContainerOutput = ContainerOutput
+> extends Embeddable<TContainerInput, TContainerOutput>
+  implements IContainer<TContainerInput, TContainerOutput> {
+  public readonly isContainer: boolean = true;
+  protected readonly children: {
+    [key: string]: IEmbeddable<any, any> | ErrorEmbeddable;
+  } = {};
+  public readonly embeddableFactories: IRegistry<EmbeddableFactory>;
+
+  private subscription: Subscription;
+
+  constructor(
+    input: TContainerInput,
+    output: TContainerOutput,
+    embeddableFactories: IRegistry<EmbeddableFactory>,
+    parent?: Container
+  ) {
+    super(input, output, parent);
+    this.embeddableFactories = embeddableFactories;
+    this.subscription = this.getInput$().subscribe(() => this.maybeUpdateChildren());
+  }
+
+  public updateInputForChild<EEI extends EmbeddableInput = EmbeddableInput>(
+    id: string,
+    changes: Partial<EEI>
+  ) {
+    if (!this.input.panels[id]) {
+      throw new Error(
+        i18n.translate('embeddableApi.errors.paneldoesNotExist', {
+          defaultMessage: 'Panel not found.',
+        })
+      );
+    }
+    const panels = {
+      panels: {
+        ...this.input.panels,
+        [id]: {
+          ...this.input.panels[id],
+          explicitInput: {
+            ...this.input.panels[id].explicitInput,
+            ...changes,
+          },
+        },
+      },
+    };
+    this.updateInput(panels as Partial<TContainerInput>);
+  }
+
+  public reload() {
+    Object.values(this.children).forEach(child => child.reload());
+  }
+
+  public async addNewEmbeddable<
+    EEI extends EmbeddableInput = EmbeddableInput,
+    EEO extends EmbeddableOutput = EmbeddableOutput,
+    E extends IEmbeddable<EEI, EEO> = IEmbeddable<EEI, EEO>
+  >(type: string, explicitInput: Partial<EEI>): Promise<E | ErrorEmbeddable> {
+    const factory = this.embeddableFactories.get(type) as
+      | EmbeddableFactory<EEI, EEO, E>
+      | undefined;
+
+    if (!factory) {
+      throw new EmbeddableFactoryNotFoundError(type);
+    }
+
+    const panelState = this.createNewPanelState<EEI, E>(factory, explicitInput);
+
+    this.updateInput({
+      panels: {
+        ...this.input.panels,
+        [panelState.embeddableId]: panelState,
+      },
+    } as Partial<TContainerInput>);
+
+    return await this.untilEmbeddableLoaded<E>(panelState.embeddableId);
+  }
+
+  public async addSavedObjectEmbeddable<
```

Comment:
It's only used by the add panel, but I like that it's part of this container class because then it can follow the same flow that adding a brand new embeddable that isn't backed by a saved object goes through.  And that is:

- Sets the container input state correctly (adds a new entry to `panels`)
- Waits for the container to recognize that input state has changed, notices a new panel has been added
- Attempts to create a new Embeddable instance from that panel data.


I could possibly not expose these methods at all and make all users of this container interface do something like

```
container.updateInput({
  panels: {
    ...container.getInput().panels,
    [newId]: newPanelState
  }
});
``` 

But by having these inside the class it encapsulates a lot of logic that callers using that simplified version would need to do (e.g. get the factory, use the factory to help create the input state, use the container as well to figure out what the input state should be... and more).

That's why it's internal, but there are technically two ways you can add and remove embeddables from a container - by changing the input state directly (like say someone goes and edits the URL in dashboard and pastes in a different visualization id, dashboard app will detect this as a change to input state and trigger the above version).

It is a shame we have to treat saved object embeddables as differently than other embeddables.  I went through multiple iterations where the saved object id was actually a part of `explicitInput` and not on the container itself but that had it's own challenges and I landed on keeping it out as a cleaner API.

Reply:
I might be able to refactor some of this since you are right, they are very similar... let me see if I can do that.  Though I still think that externally they have to be different fns because the one expects explicit input. It's the flow you go through when you add a brand new embeddable to a container (e.g. `Add new Character Card` example from the GAH session).  

We don't really use this flow yet though because the existing types all are backed by a saved object, and the `add new visualization` flow needs some help (ideally inline creation to create the explicitInput... what if we didn't even need visualizations to have their own saved object at all but all the data needed to create one was stored with the dashboard saved object?  This flat data structure has been something @skearns64 has brought up quite a bit.. I think it's because of security reasons... I'm not aware of any concrete plans to try to flatten the structure but knowing that it's possible in this model I think will help us be more flexible).